File: PixelSolution/PixelEngine/Renderer2D.cpp
#include "pch.h"
#include "Graphics.h"
#include "Renderer2D.h"
#include "PixelEngineAPI.h"
#include "Registry.h"
#include "PixelEngine.h"
#include "ResourceManager.h"
extern PixelEngine* Engine;
Renderer2DData* Renderer2D_Add(unsigned int id)
{
	auto registry = GetRegistry();
	if (registry->Has<Renderer2DData>(id) == false)
	{
		registry->Add<Renderer2DData>(id);
	}
	if (registry->Has<GraphicsData>(id) == false)
	{
		registry->Add<GraphicsData>(id);
	}
	auto graphics = registry->Get<GraphicsData>(id);
	graphics->renderingData.renderType = RENDER_TYPE::QUAD;
	graphics->renderingData.passType = PASS_TYPE::SCENE;
	graphics->renderingData.sprite.Order = 0;
	graphics->renderingData.sprite.TilingX = 1.0f;
	graphics->renderingData.sprite.TilingY = 1.0f;
	graphics->renderingData.sprite.OffsetX = 0.0f;
	graphics->renderingData.sprite.OffsetY = 0.0f;
	graphics->renderingData.sprite.color[0] = 1.0f;
	graphics->renderingData.sprite.color[1] = 1.0f;
	graphics->renderingData.sprite.color[2] = 1.0f;
	graphics->renderingData.sprite.color[3] = 1.0f;

	auto renderer2D = registry->Get<Renderer2DData>(id);
	return renderer2D;
}
Renderer2DData* Renderer2D_Get(unsigned int id)
{
	auto registry = GetRegistry();
	Renderer2DData* data = registry->Get<Renderer2DData>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Transform][GetComponent] Not Find Component");
	}
	return data;
}
bool Renderer2D_Has(unsigned int id)
{
	auto registry = GetRegistry();
	Renderer2DData* data = registry->Get<Renderer2DData>(id);
	if (data == nullptr)
	{
		return false;
	}
	else
	{
		return true;
	}
}

void Renderer2D_SetTexture(unsigned int id, const char* name)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Renderer2DData>(id);
	auto data2 = registry->Get<GraphicsData>(id);
	if (data1 != nullptr)
	{
		std::string textureName(name);
		auto textureID = Engine->GetResourceID(RESOURCE_TYPE::TEXTURE, textureName);
		data2->renderingData.texture_key = textureID;
	}
	else
	{
		PixelLog::Error("[Renderer2D][SetTexture] Not Find Component");
	}
}

void Renderer2D_SetColor(unsigned int id, float r, float g, float b, float a)
{
	auto registry = GetRegistry();
	auto renderer = registry->Get<Renderer2DData>(id);
	auto graphics = registry->Get<GraphicsData>(id);
	if (renderer == nullptr || graphics == nullptr)
	{
		PixelLog::Error("[Renderer2D][SetColor] Not Find Component");
		return;
	}

	auto normalizeColor = [](float value)
		{
			return (std::clamp)(value, 0.0f, 255.0f) / 255.0f;
		};

	graphics->renderingData.sprite.color[0] = normalizeColor(r);
	graphics->renderingData.sprite.color[1] = normalizeColor(g);
	graphics->renderingData.sprite.color[2] = normalizeColor(b);
	graphics->renderingData.sprite.color[3] = normalizeColor(a);
}

File: Survival/Engine/Renderer2D.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

ffi.cdef[[
	typedef struct {
		unsigned int thisID;
		float TilingX;
		float TilingY;
		float OffsetX;
		float OffsetY;
	} Renderer2DData;

	Renderer2DData* Renderer2D_Add(unsigned int ID);
	Renderer2DData* Renderer2D_Get(unsigned int ID);
	bool Renderer2D_Has(unsigned int ID);
	void Renderer2D_SetTexture(unsigned int ID,const char* textureName);
	void Renderer2D_SetColor(unsigned int ID,float R,float G,float B,float A);
]]

---@class Renderer2DData
---@field TilingX number
---@field TilingY number
---@field OffsetX number
---@field OffsetY number
local Renderer2DData = {}

---@param textureName string
function Renderer2DData:SetTexture(textureName)
	return dll.Renderer2D_SetTexture(self.thisID,textureName)
end

---@param R number
---@param G number
---@param B number
---@param A number
function Renderer2DData:SetColor(R,G,B,A)
	return dll.Renderer2D_SetColor(self.thisID,R,G,B,A)
end

local Renderer2D_mt = {
	__index = Renderer2DData
}

ffi.metatype("Renderer2DData", Renderer2D_mt)

Renderer2D = Renderer2D or {}

---@param ID number 
---@return Renderer2DData 
function Renderer2D.Add(ID) 
	local data = dll.Renderer2D_Add(ID)
	data.thisID = ID
	return data 
end 

---@param ID number 
---@return Renderer2DData 
function Renderer2D.Get(ID) 
	return dll.Renderer2D_Get(ID) 
end 

---@param ID number 
---@return boolean 
function Renderer2D.Has(ID) 
	return dll.Renderer2D_Has(ID) 
end 

return Renderer2D